### PR TITLE
Add reaction in a solvent option

### PR DESCRIPTION
**Pull request by yunsiechung**
This enables the Chemprop to process reaction in solvent with `--reaction_solvent` option. The input csv file should have two columns each containing the reaction and solvent SMILES. The two columns can be ordered in anyway as Chemprop can identify which column corresponds to reaction SMILES.

Differently sized MPNNs can be built for reaction and solvent. The size of the reaction MPNN can be specified with `--bias`, `--hidden_size`, and `--depth` arguments while that of the solvent MPNN can be specified with `--bias_solvent`, `--hidden_size_solvent`, and `--depth_solvent`.

The README.md file has been updated to add `Reaction in a solvent` section.

### CODE REVIEW

**Comment by hesther** [comment]
I think we broke `get_invalid_smiles_from_list()` in chemprop/data/utils.py . Since make_mols now excepts lists of `reaction`, `keep_h` and `add_h`, we need to adapt it in this function, too. (This is causing the LGTM alert).

**Comment by hesther** [comment]
Apart from the three comments above (small changes in README and featurization.py, bugfix in utils.py) I think this is good. We should also introduce a test + a test set in tests/data.

**Comment by yunsiechung** [comment]
@hesther Thanks for the comment. I will try to fix the bug with `get_invalid_smiles_from_list`.

**Comment by yunsiechung** [comment]
The bug in utils.py has been fixed and the test and test data for reaction_solvent have been added to `tests/test_integration.py` and `tests/data/reaction_solvent_regression.csv`.

File: chemprop/features/featurization.py [hunk 1]
```diff
@@ -281,6 +299,11 @@ class MolGraph:
     * :code:`b2revb`: A mapping from a bond index to the index of the reverse bond.
     * :code:`overwrite_default_atom_features`: A boolean to overwrite default atom descriptors.
     * :code:`overwrite_default_bond_features`: A boolean to overwrite default bond descriptors.
+    * :code:`is_mol`: A boolean whether the input is a molecule.
+    * :code:`is_reaction`: A boolean whether the molecule is a reaction.
+    * :code:`is_explicit_h`: A boolean whether to retain explicit Hs (for reaction mode)
+    * :code:`is_explicit_h`: A boolean whether to add explicit Hs (not for reaction mode)
```
Comment (shihchengli) [inline chemprop/features/featurization.py:305]: The `is_explicit_h` in line 305 should be `is_adding_hs`?

Comment (yunsiechung) [reply chemprop/features/featurization.py]: Thank you for catching it! I made a fixup commit [51d0e81](https://github.com/chemprop/chemprop/pull/246/commits/51d0e8132eae154c15bc05d67023b47dd0e4cd80) to correct this typo.

**Comment by hesther** [comment]
Am restarting the unit tests since one of them timed out. Yunsie do you want to make the two changes (in README and featurization.py) from my comments above (or I can push them anytime)? I also thought that the way I set up two is_mol functions in featurization.py and data.py is a bit clumsy, and wanted to unify them to one

**Comment by yunsiechung** [comment]
@hesther I cannot find your comment on the two changes (in README and featurization.py). Where can I find it? Can you tag me in that comment?

**Comment by hesther** [comment]
@yunsiechung , don't you see this when you scroll up the comments here:
test

**Comment by hesther** [comment]
@yunsiechung , oh, my bad, somehow the comments I wrote last week were stuck (I saw them in my feed, but I think no-one else saw them... now they should be visible!)

File: chemprop/features/featurization.py [hunk 2]
```diff
@@ -45,6 +45,7 @@ def __init__(self) -> None:
         self.REACTION_MODE = None
         self.EXPLICIT_H = False
         self.REACTION = False
+        self.REACTION_SOLVENT = False
```
[the remark this answers is not in the document]
Comment (yunsiechung) [reply chemprop/features/featurization.py]: Thanks for catching it. I deleted this line with this fixup commit [ad96503](https://github.com/chemprop/chemprop/pull/246/commits/ad96503fcd5ef488d89ed2e1cb301236a996d8e3).

File: README.md [hunk 3]
```diff
@@ -269,6 +270,33 @@ In absorption spectra, sometimes the phase of collection will create regions in
...
 For further details and benchmarking, as well as a citable reference, please refer to the [article](https://doi.org/10.1021/acs.jcim.1c00975).
 
+### Reaction in a solvent
+
+Chemprop can process reaction in a solvent with the `--reaction_solvent` option.
```
[the remark this answers is not in the document]
Comment (yunsiechung) [reply README.md]: That's a very good point! I added that to the README file in this fixup commit [da12957](https://github.com/chemprop/chemprop/pull/246/commits/da12957c1d22e10b2ce1a108f461a3e95582d378).

**Comment by hesther** [comment]
I just cleaned up the is_mol function as written above. @yunsiechung I think this is ready to be squashed

**Review by hesther — APPROVED** [review]

**Comment by yunsiechung** [comment]
@hesther Thanks! I squashed the fixup commits and rebased the current branch with the latest `master` branch. It should be ready to be merged.

**Comment by hesther** [comment]
In the past, we usually tried to squash the commits down to one or two commits to keep the history easier to read. But I am also happy with more commits, if no-one else disagrees

**Comment by hesther** [comment]
@yunsiechung Whatever you and the other new team members like best - it is really just a personal style question, and the old team liked less commits better, but you are totally free to choose what you like best!

**Comment by cjmcgill** [comment]
I'm a fan of fewer commits, but that doesn't have to mean just 1 commit. If you have like reasonable groupings of them you could squash into a few commits.

**Comment by yunsiechung** [comment]
@hesther @cjmcgill Thanks for the comments. I squashed them to just one commit because there are all related to each other. Let me know if I didn't squash them properly. Otherwise, it should be ready to be merged.

**Comment by hesther** [comment]
Looks good to me! Let's wait for the tests and then merge

**Pull request merged**